nbconvert/preprocessors/tagremove.py

```diff
@@ -0,0 +1,101 @@
+        """
+        Checks that a cell has a tag that is to be removed
+
+        Returns: Boolean.
+        True means cell should *not* be removed.
+        """
+
+        # Return true if any of the tags in the cell are removable.
+        return not any([tag in cell.get('metadata', {}).get('tags', [])
+                        for tag in self.remove_cell_tags])
+
+    def preprocess(self, nb, resources):
+        """
+        Preprocessing to apply to each notebook. See base.py for details.
+        """
+        # Skip preprocessing if the list of patterns is empty
+        if not any([self.remove_cell_tags,
+                    self.remove_all_outputs_tags,
+                    self.remove_single_output_tags]):
+            return nb, resources
+
+        # Filter out cells that meet the conditions
+        nb.cells = [self.preprocess_cell(cell, resources, index)[0]
+                    for index, cell in enumerate(nb.cells)
+                    if self.check_cell_conditions(cell, resources, index)]
+
+        return nb, resources
+
+    def preprocess_cell(self, cell, resources, cell_index):
+        """
+        Apply a transformation on each cell. See base.py for details.
+        """
+
+        if (any([tag in cell.get('metadata', {}).get('tags', [])
+                 for tag in self.remove_all_outputs_tags])
+            and cell.cell_type == 'code'):
+            cell.outputs = []
+            cell.execution_count = None
+            # Remove metadata associated with output
+            if 'metadata' in cell:
+                for field in self.remove_metadata_fields:
+                    cell.metadata.pop(field, None)
+        if cell.get('outputs', {}):
+            cell.outputs = [self.preprocess_output(output,
+                                                   resources,
+                                                   cell_index,
+                                                   output_index)[0]
+                            for output_index, output in enumerate(cell.outputs)
+                            if self.check_output_conditions(output,
+                                                            resources,
+                                                            cell_index,
+                                                            output_index)
+                            ]
+        return cell, resources
+
+    def check_output_conditions(self, output, resources,
+                                cell_index, output_index):
+        """
+        Checks that an output has a tag that indicates removal.
+
+        Returns: Boolean.
+        True means output should *not* be removed.
+        """
+        return not any([tag in output.get('metadata', {}).get('tags', [])
+                        for tag in self.remove_single_output_tags])
+
+    def preprocess_output(self, output, resources,
+                          cell_index, output_index):
+        return output, resources
```

Comment:
I was partially trying to give people a handle on preprocessing individual outputs if they wanted to subclass (I thought it might be helpful for things like tag based preprocessing… but I'm realising now that something like that is more broad ranging and should probably be put into a different PR).